File: demo/Blackholio/client-unity/Assets/Scripts/PlayerController.cs
using System.Collections.Generic;
using System.Linq;
using SpacetimeDB;
using SpacetimeDB.Types;
using UnityEngine;

public class PlayerController : MonoBehaviour
{
    const int SEND_UPDATES_PER_SEC = 20;
    const float SEND_UPDATES_FREQUENCY = 1f / SEND_UPDATES_PER_SEC;

    public static PlayerController Local { get; private set; }

    private int PlayerId;
    private float LastMovementSendTimestamp;
    private Vector2? LockInputPosition;
    private List<CircleController> OwnedCircles = new List<CircleController>();

    public string Username => GameManager.Conn.Db.Player.PlayerId.Find(PlayerId).Name;
    public int NumberOfOwnedCircles => OwnedCircles.Count;
    public bool IsLocalPlayer => this == Local;

    public void Initialize(Player player)
    {
        PlayerId = player.PlayerId;
        if (player.Identity == GameManager.LocalIdentity)
        {
            Local = this;
        }
    }

    private void OnDestroy()
    {
        // If we have any circles, destroy them
        foreach (var circle in OwnedCircles)
        {
            if (circle != null)
            {
                Destroy(circle.gameObject);
            }
        }
        OwnedCircles.Clear();
    }

    public void OnCircleSpawned(CircleController circle)
    {
        OwnedCircles.Add(circle);
    }

    public void OnCircleDeleted(CircleController deletedCircle)
    {
        // This means we got eaten
        if (OwnedCircles.Remove(deletedCircle) && IsLocalPlayer && OwnedCircles.Count == 0)
        {
            GameManager.Instance.deathScreen.SetVisible(true);
        }
    }

    public int TotalMass()
    {
        return (int)OwnedCircles
            .Select(circle => GameManager.Conn.Db.Entity.EntityId.Find(circle.EntityId))
            .Sum(e => e?.Mass ?? 0); //If this entity is being deleted on the same frame that we're moving, we can have a null entity here.
    }

    public Vector2? CenterOfMass()
    {
        if (OwnedCircles.Count == 0)
        {
            return null;
        }

        Vector2 totalPos = Vector2.zero;
        float totalMass = 0;
        foreach (var circle in OwnedCircles)
        {
            var entity = GameManager.Conn.Db.Entity.EntityId.Find(circle.EntityId);
            var position = circle.transform.position;
            totalPos += (Vector2)position * entity.Mass;
            totalMass += entity.Mass;
        }

        return totalPos / totalMass;
    }

    public void Update()
    {
        if (!IsLocalPlayer || NumberOfOwnedCircles == 0)
        {
            return;
        }

        if (Input.GetKeyDown(KeyCode.Space))
        {
            GameManager.Conn.Reducers.PlayerSplit();
        }

        if (Input.GetKeyDown(KeyCode.Q))
        {
            if (LockInputPosition.HasValue)
            {
                LockInputPosition = null;
            }
            else
            {
                LockInputPosition = (Vector2)Input.mousePosition;
            }
        }

        if (Input.GetKeyDown(KeyCode.S))
        {
            GameManager.Conn.Reducers.Suicide();
        }

        // Throttled input requests
        if (Time.time - LastMovementSendTimestamp >= SEND_UPDATES_FREQUENCY)
        {
            LastMovementSendTimestamp = Time.time;

            var mousePosition = LockInputPosition ?? (Vector2)Input.mousePosition;
            var screenSize = new Vector2
            {
                x = Screen.width,
                y = Screen.height,
            };
            var centerOfScreen = screenSize / 2;

            var direction = (mousePosition - centerOfScreen) / (screenSize.y / 3);
            if (testInputEnabled) { direction = testInput; }
            GameManager.Conn.Reducers.UpdatePlayerInput(direction);
        }
    }

    private void OnGUI()
    {
        if (!IsLocalPlayer || !GameManager.IsConnected())
        {
            return;
        }

        GUI.Label(new Rect(0, 0, 100, 50), $"Total Mass: {TotalMass()}");
    }

    //Automated testing members
    private bool testInputEnabled;
    private Vector2 testInput;

    public void SetTestInput(Vector2 input) => testInput = input;
    public void EnableTestInput() => testInputEnabled = true;
}

File: demo/Blackholio/client-ts/src/game/PlayerController.ts
import Phaser from 'phaser';
import type { Player } from '../module_bindings/types';
import { pointerDirection } from './input';
import { centerOfMass, type Vec2, type WeightedPosition } from './math';
import type { CircleController } from './CircleController';
import type { GameManager } from './GameManager';

const SEND_UPDATES_FREQUENCY_MS = 1000 / 20;

export class PlayerController {
  private readonly ownedCircles: CircleController[] = [];
  private readonly splitKey?: Phaser.Input.Keyboard.Key;
  private readonly lockKey?: Phaser.Input.Keyboard.Key;
  private readonly suicideKey?: Phaser.Input.Keyboard.Key;
  private lastMovementSendTimestamp = 0;
  private lockedPointer?: Vec2;

  constructor(
    private readonly gameManager: GameManager,
    private player: Player
  ) {
    const keyboard = gameManager.scene.input.keyboard;
    this.splitKey = keyboard?.addKey(Phaser.Input.Keyboard.KeyCodes.SPACE, false);
    this.lockKey = keyboard?.addKey(Phaser.Input.Keyboard.KeyCodes.Q, false);
    this.suicideKey = keyboard?.addKey(Phaser.Input.Keyboard.KeyCodes.S, false);
  }

  get playerId(): number {
    return this.player.playerId;
  }

  get username(): string {
    return this.player.name;
  }

  get isLocalPlayer(): boolean {
    return this.gameManager.identity?.isEqual(this.player.identity) ?? false;
  }

  get numberOfOwnedCircles(): number {
    return this.ownedCircles.length;
  }

  updatePlayer(player: Player): void {
    this.player = player;
    this.ownedCircles.forEach(circle => circle.updateUsername());
  }

  onCircleSpawned(circle: CircleController): void {
    this.ownedCircles.push(circle);
  }

  onCircleDeleted(entityId: number): void {
    const index = this.ownedCircles.findIndex(circle => circle.entityId === entityId);
    if (index !== -1) {
      this.ownedCircles.splice(index, 1);
      if (this.isLocalPlayer && this.ownedCircles.length === 0) {
        this.gameManager.deathScreen.setVisible(true);
      }
    }
  }

  totalMass(): number {
    return this.ownedCircles.reduce(
      (sum, circle) => sum + (this.gameManager.findEntity(circle.entityId)?.mass ?? 0),
      0
    );
  }

  centerOfMass(): Vec2 | undefined {
    const entities: WeightedPosition[] = this.ownedCircles.flatMap(circle => {
      const entity = this.gameManager.findEntity(circle.entityId);
      return entity ? [{ mass: entity.mass, position: circle.position }] : [];
    });
    return centerOfMass(entities);
  }

  update(time: number): void {
    if (!this.isLocalPlayer || this.ownedCircles.length === 0) {
      return;
    }
    if (this.splitKey && Phaser.Input.Keyboard.JustDown(this.splitKey)) {
      void this.gameManager.connection?.reducers.playerSplit({});
    }
    if (this.lockKey && Phaser.Input.Keyboard.JustDown(this.lockKey)) {
      this.lockedPointer = this.lockedPointer
        ? undefined
        : this.currentPointerPosition();
    }
    if (this.suicideKey && Phaser.Input.Keyboard.JustDown(this.suicideKey)) {
      void this.gameManager.connection?.reducers.suicide({});
    }
    if (time - this.lastMovementSendTimestamp < SEND_UPDATES_FREQUENCY_MS) {
      return;
    }
    this.lastMovementSendTimestamp = time;
    const direction = pointerDirection(
      this.lockedPointer ?? this.currentPointerPosition(),
      {
        x: this.gameManager.scene.scale.width,
        y: this.gameManager.scene.scale.height,
      }
    );
    void this.gameManager.connection?.reducers.updatePlayerInput({ direction });
  }

  private currentPointerPosition(): Vec2 {
    const pointer = this.gameManager.scene.input.activePointer;
    return { x: pointer.x, y: pointer.y };
  }
}
